thinc/optimizers.py

```diff
@@ -354,4 +354,127 @@ def _adam(self, ops, weights, gradient, lr_scale, key, nr_upd):
         )
 
 
+class SWA(object):
+    """
+    After the optimization ran 'start_step' number of steps
+    it switches the learning-rate of the optimizer and starts
+    recording the moving averages of all weights.
+    """
+
+    def __init__(
+        self,
+        optimizer: Optimizer,
+        lr: FloatOrSeq,
+        *,
+        freq: int = 1,
+        start_step: Optional[int] = 0
+    ):
+        self.optimizer = optimizer
+        self.start_step = start_step
+        self.freq = freq
+        self.lr = lr
+        self.nr_update: Dict[Tuple[int, str], int] = defaultdict(int)
+        self.run_swa = False
+        self.averages: Dict[Tuple[int, str], FloatsXd] = {}
+
+    def _update_averages(
+        self,
+        key: Tuple[int, str],
+        weights: FloatsXd,
+    ):
+        """
+        Update moving average of weights.
+        """
+        if key not in self.averages:
+            ops = get_array_ops(weights)
+            self.averages[key] = ops.alloc(weights.shape, dtype="float32")
+            self.averages[key] += weights
+        else:
+            self.averages[key] *= self.nr_update[key]
+            self.averages[key] += weights
+            self.averages[key] /= self.nr_update[key] + 1
+
+    def _swap_lr(self):
+        """
+        Overwrite the optimizer's learning rate.
+        """
+        self.optimizer._set_attr_or_schedule("learn_rate", self.lr)
+
+    def start_swa(self):
+        """
+        Instead of providing 'start_step' a caller
+        can turn SWA on with this method.
+        """
+        self.run_swa = True
+        self._swap_lr()
+
+    def __call__(
+        self,
+        key: Tuple[int, str],
+        weights: FloatsXd,
+        gradient: FloatsXd,
+        *,
+        lr_scale: float = 1.0,
+    ):
+        self.nr_update[key] += 1
+        nr_update = self.nr_update[key]
+        weights, gradient = self.optimizer(key, weights, gradient, lr_scale=lr_scale)
+        # Update running mean
+        if nr_update == self.start_step:
```

Comment:
I am not sure if I understand this. SWA is triggered once one of the variables reaches the given number of updates. Wouldn't we want to decide this per variable? If so, then I think changing the condition below to:

```
        if nr_update >= self.start_step and nr_update % self.freq == 0:
             self._update_averages(key, weights)
```

Also removes some state from the class (`self.run_swa`).